cmd/kritis/signer/server.go

```diff
@@ -0,0 +1,228 @@
+	w.WriteHeader(code)
+	w.Write(body)
+}
+
+// Google Pub/Sub message
+type PubSubMessage struct {
+	Message struct {
+		Data []byte `json:"data,omitempty"`
+		ID   string `json:"id"`
+	} `json:"message"`
+	Subscription string `json:"subscription"`
+}
+
+// Google Container Analysis occurrence notification
+type ContainerAnalysisOccurrenceEvent struct {
+	Name             string    `json:"name"`
+	Kind             string    `json:"kind"`
+	NotificationTime time.Time `json:"notificationTime"`
+}
+
+// retrieve the container analysis occurrence associated with this notification
+func GetOccurrence(event ContainerAnalysisOccurrenceEvent) (*grafeaspb.Occurrence, error) {
+	request := grafeaspb.GetOccurrenceRequest{Name: event.Name}
+	response, err := grafeas.GetOccurrence(context.Background(), &request)
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve discovery occurrence %s", event.Name)
+	}
+	return response, nil
+}
+
+// implements `/event` handles a container analysis occurrence notification
+func containerAnalysisEvent(w http.ResponseWriter, r *http.Request) {
+	var m PubSubMessage
+	var event ContainerAnalysisOccurrenceEvent
+	err := json.NewDecoder(r.Body).Decode(&m)
+	if err != nil {
+		http.Error(w, "could not unmarshal the pubsub message", http.StatusBadRequest)
+		return
+	}
+
+	err = json.Unmarshal(m.Message.Data, &event)
+	if err != nil {
+		http.Error(w, "could not unmarshal container analysis occurrence event from pubsub message", http.StatusBadRequest)
+		return
+	}
+
+	if event.Kind != "DISCOVERY" {
+		http.Error(w, "ignoring non-DISCOVERY occurrence", http.StatusOK)
+		return
+	}
+
+	occurrence, err := GetOccurrence(event)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	status := occurrence.GetDiscovered().GetDiscovered().GetAnalysisStatus()
+	if status != discovery.Discovered_FINISHED_SUCCESS {
+		http.Error(w, fmt.Sprintf("ignoring DISCOVERY occurrence in status %s", string(status)), http.StatusOK)
+		return
+	}
+	image := occurrence.Resource.Uri
+	if strings.HasPrefix(image, "https://") {
+		image = image[8:]
+	}
+	request := SignRequest{Image: image}
+	doCheckAndSign(w, request, http.StatusOK)
+}
+
+// implements `/check-only`
+func check(w http.ResponseWriter, r *http.Request) {
+	var request SignRequest
+	var response SignResponse
+
+	err := json.NewDecoder(r.Body).Decode(&request)
+	if err != nil {
+		response.Status = Failed
+		response.Message = fmt.Sprintf("failed to decode body, %s", err)
+		WriteResponse(w, response, http.StatusBadRequest)
+		return
+	}
+	response.Image = request.Image
+
+	violations, err := DoCheck(request.Image)
+	if err != nil {
+		response.Status = Failed
+		response.Message = err.Error()
+		WriteResponse(w, response, http.StatusBadRequest)
+		return
+	}
+	if violations != nil {
+		response.Status = Failed
+		response.Violations = violations
+		WriteResponse(w, response, http.StatusUnprocessableEntity)
```

Comment:
StatusBadRequest, like in above cases?

Reply:
StatusBadRequest (400) indicates that the input is incorrect and the caller should fix it.
StatusUnprocessableEntity means the input is correct, but we are  unable to process the contained instructions. 

We could change it to 200 (as we are requested to perform a check, which we did) and have the response checked for success or failure. But then it would be a little bit harder to use this /check call in a program flow (for instance curl).